### PR TITLE
More models

### PR DESCRIPTION
**Pull request by xiaolong0728**
I am making a new pull request because I forgot to merge the previous one but pushed something later :(

I've tested the codes on two stepshifter models - **lavender haze** (hurdle model) and **orange pasta** (ml model). So you can temporarily ignore those changes I made weeks ago to other models.

Some improvements I have noted down:
1. Split functions utils.py into different files 
2. A lot of compromises are made for hurdle models. There should be a better way to adapt to it.
3. Align eval functions with Simon's. (in utils_model_outputs.py and utils_evaluation_metrics.py)

### CODE REVIEW

File: models/blank_space/src/training/train_model.py [hunk 1]
```diff
@@ -16,45 +16,37 @@
...
-        PATH_ARTIFACTS = setup_artifacts_paths(PATH)
-        dataset = pd.read_parquet(PATH_RAW / 'raw.parquet')
-        if model_config["algorithm"] == "HurdleRegression":
-            model = HurdleRegression(clf_name=model_config["clf_name"], reg_name=model_config["reg_name"], clf_params=para_config["clf"], reg_params=para_config["reg"])
-        else:
-            model = globals()[model_config["algorithm"]](**para_config)
-        # print(model)
-            
-        # Train partition
+def get_model(model_config, para_config):
```
Comment (Polichinel) [inline models/blank_space/src/training/train_model.py:19]: Do we have a config called para_config? It's the config_hyperparameters right? If so, just use the full name here

File: models/electric_relaxation/src/utils/set_paths.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ def get_artifacts_path(partition_name):
 
 def get_data_path(data_name):
     '''
-    E.g., The data is saved in data/raw/raw.parquet
+    E.g., The data is saved in data/raw/raw_calibration.parquet
```
Comment (Polichinel) [inline models/electric_relaxation/src/utils/set_paths.py:17]: I'm not sure we need to denote it "raw_" when it is in the raw folder, but we haven't agreed on any conventions yet. But I would probably go with just "calibration.parquet"

File: models/electric_relaxation/src/utils/set_paths.py [hunk 3]
```diff
@@ -14,7 +14,7 @@ def get_artifacts_path(partition_name):
...
     '''
 
     return Path(__file__).parent.parent.parent / "data" / f"{data_name}" / f"{data_name}.parquet"
```
Comment (Polichinel) [inline models/electric_relaxation/src/utils/set_paths.py:20]: Can I get you to adopt this solution instead:

PATH = Path(__file__)
sys.path.insert(0, str(Path(*[i for i in PATH.parts[:PATH.parts.index("views_pipeline")+1]]) / "common_utils")) 
from set_path import setup_project_paths, setup_data_paths
setup_project_paths(PATH)

PATH_RAW, PATH_PROCESSED, PATH_GENERATED = setup_data_paths(PATH)

File: models/lavender_haze/configs/config_hyperparameters.py [hunk 4]
```diff
@@ -1,14 +1,22 @@
...
-        "clf":{
-            "learning_rate": 0.05,
-            "n_estimators": 100,
-            "n_jobs": 12
-        },
-        "reg":{
-            "learning_rate": 0.05,
-            "n_estimators": 100,
-            "n_jobs": 12
+        "name": "lavender_haze",
```
Comment (Polichinel) [inline models/lavender_haze/configs/config_hyperparameters.py:3]: We need to think here a bit. Because the config_hyperparameters is meant to only contain "operational" parameters. I.e. parameters you can actually change here to change to model. And I'm not sure that is the case for:

       "algorithm": "HurdleRegression",
        "model_clf": "LGBMClassifier",
        "model_reg": "LGBMRegressor",
        "depvar": "ln_ged_sb_dep",
        
I mean, could I change any of these just here, and then get a different but still functional model?

I think I have produced some writing somewhere about it, but given what we talked about yesterday I'll make that more clear.

File: models/orange_pasta/configs/config_hyperparameters.py [hunk 5]
```diff
@@ -1,7 +1,13 @@
 def get_hp_config(): 
     hp_config = {
-        "learning_rate": 0.05,
-        "n_estimators": 100,
-        "n_jobs": 12   
+        "name": "orange_pasta",
+        "algorithm": "LGBMRegressor",
+        "depvar": "ged_sb_dep",
```
Comment (Polichinel) [inline models/orange_pasta/configs/config_hyperparameters.py:5]: Again, these do not seem operational and might below in he config_meta

File: models/orange_pasta/notebooks/common_utils/utils_model_outputs.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        Note:
+            - List-like columns are expanded such that each element in the list appears in a new row.
+            - Attributes in ModelOutputs that are not lists remain as-is in the DataFrame.
+        """ 
+
+        # Convert directly to a DataFrame and then explode list-like columns
+        df = pd.DataFrame([{attr: getattr(instance, attr) for attr in instance.__dataclass_fields__.keys()} for instance in dict_of_outputs.values()]).apply(pd.Series.explode)
+
+        return df
```
Comment (Polichinel) [inline models/orange_pasta/notebooks/common_utils/utils_model_outputs.py:109]: Again, the conflict seems to be just a comment. No biggy

**Review by Polichinel — CHANGES_REQUESTED** [review]
Cool! Looks good. A few things to do, but minimal. Also the two conflicts seem to be just comments I put, where you then put code. Just move the comments to the end of the script or delete if out-dated. Should be simple

File: models/electric_relaxation/src/utils/set_paths.py (hunk 2 above)
Comment (xiaolong0728) [reply models/electric_relaxation/src/utils/set_paths.py:17]: This is Sara's model. I haven't updated it yet.

File: models/electric_relaxation/src/utils/set_paths.py (hunk 3 above)
Comment (xiaolong0728) [reply models/electric_relaxation/src/utils/set_paths.py:20]: Same above

File: models/lavender_haze/configs/config_hyperparameters.py (hunk 4 above)
Comment (xiaolong0728) [reply models/lavender_haze/configs/config_hyperparameters.py:3]: Yes, changing these parameters will produce a different model.

File: models/orange_pasta/configs/config_hyperparameters.py (hunk 5 above)
Comment (xiaolong0728) [reply models/orange_pasta/configs/config_hyperparameters.py:5]: I actually need these parameters to generalize codes, so changing them will affect the process. For example, the name is used to initialize the project name; the algorithm is used to initialize a model. Otherwise I have to hardcode them.

File: common_utils/utils_model_outputs.py [hunk 7]
```diff
@@ -22,7 +22,7 @@ class ModelOutputs:
         pg_id (Optional[List[int]]): The priogrid id.
         c_id (Optional[List[int]]): The country id.
         month_id (Optional[List[int]]): The month id.
-        step (Optional[List[int]]): The step ahead forecast.
```
[the remark this answers is not in the document]
Comment (xiaolong0728) [reply common_utils/utils_model_outputs.py:25]: Noted! I will implement these changes accordingly.

File: common_utils/utils_evaluation_metrics.py [hunk 8]
```diff
@@ -77,3 +83,68 @@ def evaluation_dict_to_dataframe(evaluation_dict: dict) -> pd.DataFrame:
...
+    for step in config.steps:
+        evaluation_dict[f"step{str(step).zfill(2)}"].MSE = mean_squared_error(df[config.depvar], df[f"step_pred_{step}"])
+        evaluation_dict[f"step{str(step).zfill(2)}"].MAE = mean_absolute_error(df[config.depvar], df[f"step_pred_{step}"])
+        # evaluation_dict[f"step{str(step).zfill(2)}"].MSLE = mean_squared_log_error(df[config.depvar], df[f"step_pred_{step}"])
+        evaluation_dict[f"step{str(step).zfill(2)}"].CRPS = ps.crps_ensemble(df[config.depvar], df[f"step_pred_{step}"]).mean()
+        # evaluation_dict[f"step{str(step).zfill(2)}"].Brier = brier_score_loss(df[config.depvar], df[f"step_pred_{step}"])
+        # evaluation_dict[f"step{str(step).zfill(2)}"].AUC = roc_auc_score(df[config.depvar], df[f"step_pred_{step}"])
+        # evaluation_dict[f"step{str(step).zfill(2)}"].AP = average_precision_score(df[config.depvar], df[f"step_pred_{step}"])
+    evaluation_dict = EvaluationMetrics.output_metrics(evaluation_dict)
+    df_evaluation_dict = EvaluationMetrics.evaluation_dict_to_dataframe(evaluation_dict)
```
[the remark this answers is not in the document]
Comment (xiaolong0728) [reply common_utils/utils_evaluation_metrics.py:149]: As we discussed before, I implemented this function to create a dataframe with mean, std and median for each step under each metric. It can also be used for other kinds of models. If we no longer need it, we can stick to the original dataclass.

The config file is config_hyperparameter. I will modify it to make it clear.

File: models/electric_relaxation/src/utils/set_paths.py (hunk 2 above)
Comment (Polichinel) [reply models/electric_relaxation/src/utils/set_paths.py:17]: Noted

File: common_utils/utils_model_outputs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Polichinel) [reply common_utils/utils_model_outputs.py:25]: Let's discuss this in our meeting

File: models/electric_relaxation/src/utils/set_paths.py (hunk 3 above)
Comment (Polichinel) [reply models/electric_relaxation/src/utils/set_paths.py:20]: Super - yes, I also saw you had the "right" solution elsewhere

File: models/orange_pasta/configs/config_hyperparameters.py (hunk 5 above)
Comment (Polichinel) [reply models/orange_pasta/configs/config_hyperparameters.py:5]: Yeah, no hard coding of course. So I wrote a whole thing about this in slack at some point, but since I didn't it down somewhere more official I get why it might have been missed. But I have now put it here:

https://github.com/prio-data/views_pipeline/blob/main/documentation/ADRs/002_seperation_of_configs.md

A new documentation sub-folder mostly for internal use. This readme should explain why:

https://github.com/prio-data/views_pipeline/blob/main/documentation/ADRs/README.md

File: common_utils/utils_evaluation_metrics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Polichinel) [reply common_utils/utils_evaluation_metrics.py:149]: Sounds good

File: models/orange_pasta/configs/config_hyperparameters.py (hunk 5 above)
Comment (Polichinel) [reply models/orange_pasta/configs/config_hyperparameters.py:5]: Also, I can see that three of these are already in the meta config (where I think they below):

- "name": "orange_pasta", (already in meta)
  "algorithm": "LGBMRegressor", (already in meta)
  "depvar": "ged_sb_dep", (already in meta)
  
So I really just ask that you take them from there and not the other config (according to the ADR). 

I'm not sure about "steps": [*range(1, 36 + 1, 1)]. I have a similar parameter for Hydranet in the hyperparameters and sweep config... I'm not sure how I feel about that but I think we can both leave it for now.

**Review by Polichinel — APPROVED** [review]
